control/descfcn.py

```diff
@@ -0,0 +1,482 @@
+
+    zero_check : bool, optional
+        If `True` (default) then `A` is zero, the function will be evaluated
+        and checked to make sure it is zero.  If not, a `TypeError` exception
+        is raised.  If zero_check is `False`, no check is made on the value of
+        the function at zero.
+
+    try_method : bool, optional
+        If `True` (default), check the `F` argument to see if it is an object
+        with a `describing_function` method and use this to compute the
+        describing function.  More information in the `describing_function`
+        method for the :class:`~control.DescribingFunctionNonlinearity` class.
+
+    Returns
+    -------
+    df : complex or array of complex
+        The (complex) value of the describing function at the given amplitude.
+        If the `A` parameter is an array of amplitudes, then an array of
+        corresponding describing function values is returned.
+
+    Raises
+    ------
+    TypeError
+        If A < 0 or if A = 0 and the function F(0) is non-zero.
+
+    """
+    # If there is an analytical solution, trying using that first
+    if try_method and hasattr(F, 'describing_function'):
+        # Go through all of the amplitudes we were given
+        df = []
+        for a in np.atleast_1d(A):
+            df.append(F.describing_function(a))
+        return np.array(df).reshape(np.shape(A))
+
+    #
+    # The describing function of a nonlinear function F() can be computed by
+    # evaluating the nonlinearity over a sinusoid.  The Fourier series for a
+    # static nonlinear function evaluated on a sinusoid can be written as
+    #
+    # F(A\sin\omega t) = \sum_{k=1}^\infty M_k(A) \sin(k\omega t + \phi_k(A))
+    #
+    # The describing function is given by the complex number
+    #
+    #    N(A) = M_1(A) e^{j \phi_1(A)} / A
+    #
+    # To compute this, we compute F(\theta) for \theta between 0 and 2 \pi,
+    # use the identities
+    #
+    #   \sin(\theta + \phi) = \sin\theta \cos\phi + \cos\theta \sin\phi
+    #   \int_0^{2\pi} \sin^2 \theta d\theta = \pi
+    #   \int_0^{2\pi} \cos^2 \theta d\theta = \pi
+    #
+    # and then integrate the product against \sin\theta and \cos\theta to obtain
+    #
+    #   \int_0^{2\pi} F(a\sin\theta) \sin\theta d\theta = M_1 \pi \cos\phi
+    #   \int_0^{2\pi} F(a\sin\theta) \cos\theta d\theta = M_1 \pi \sin\phi
+    #
+    # From these we can compute M1 and \phi.
+    #
+
+    # Evaluate over a full range of angles
+    theta = np.linspace(0, 2*np.pi, num_points)
+    dtheta = theta[1] - theta[0]
+    sin_theta = np.sin(theta)
+    cos_theta = np.cos(theta)
+
+    # See if this is a static nonlinearity (assume not, just in case)
+    if not hasattr(F, '_isstatic') or not F._isstatic():
+        # Initialize any internal state by going through an initial cycle
+        [F(x) for x in np.atleast_1d(A).min() * sin_theta]
+
+    # Go through all of the amplitudes we were given
+    df = []
```

Comment:
It would be better to preallocate `df` with `np.empty`; maybe
```
    retdf = np.empty(np.shape(A), dtype=complex)
    df = retdf
    df.shape = (-1,)
```
the for loop would be `for i, a in enumerate(np.atleast_1d(A)):`, assignments would be to `df[i]`, and the function would return `retdf`.